cxx-qt-build/src/lib.rs

```diff
@@ -143,139 +144,146 @@ fn extract_modules(file_content: &str, rs_path: &str) -> ExtractedModule {
-            _others => panic!(
-                "No module to generate cxx code from could be found in {}",
-                rs_path
-            ),
+        };
+
+        let opt = cxx_gen::Opt::default();
+        let cxx = cxx_gen::generate_header_and_cc(tokens, &opt)
+            .expect("Could not generate C++ from Rust file");
+
+        GeneratedCpp {
+            cxx_qt,
+            cxx,
+            module_ident,
         }
-    };
+    }
 
-    let opt = cxx_gen::Opt::default();
-    let gen_result = cxx_gen::generate_header_and_cc(tokens, &opt)
-        .expect("Could not generate C++ from Rust file");
+    /// Write generated code to files in a directory. Returns the absolute paths of all files written.
+    pub fn write_to_directory(&self, directory: &impl AsRef<std::path::Path>) -> Vec<PathBuf> {
+        let directory = directory.as_ref();
+        if !directory.is_dir() {
+            panic!(
+                "Output directory {} is not a directory",
+                directory.display()
+            );
+        }
 
-    let mut header = File::create(&h_path).expect("Could not create header file");
-    header
-        .write_all(&gen_result.header)
-        .expect("Could not write header file");
+        let include_directory_path = PathBuf::from(format!("{}/include", &directory.display()));
+        std::fs::create_dir_all(&include_directory_path)
+            .expect("Could not create cxx-qt include dir");
+
+        let source_directory_path = PathBuf::from(format!("{}/src", &directory.display()));
+        std::fs::create_dir_all(&source_directory_path)
+            .expect("Could not create cxx-qt source dir");
+
+        let mut written_files = Vec::with_capacity(4);
+
+        if let Some(cxx_qt_generated) = &self.cxx_qt {
+            let header_path = PathBuf::from(format!(
+                "{}/{}.h",
+                include_directory_path.display(),
+                self.module_ident
+            ));
+            let mut header =
+                File::create(&header_path).expect("Could not create cxx-qt header file");
+            header
+                .write_all(cxx_qt_generated.header.as_bytes())
+                .expect("Could not write cxx-qt header file");
+            written_files.push(header_path);
+
+            let cpp_path = PathBuf::from(format!(
+                "{}/{}.cpp",
+                source_directory_path.display(),
+                &self.module_ident
+            ));
+            let mut cpp = File::create(&cpp_path).expect("Could not create cxx-qt source file");
+            cpp.write_all(cxx_qt_generated.source.as_bytes())
+                .expect("Could not write cxx-qt source file");
+            written_files.push(cpp_path);
+        }
 
-    let mut cpp = File::create(&cpp_path).expect("Could not create cpp file");
-    cpp.write_all(&gen_result.implementation)
-        .expect("Could not write cpp file");
+        let header_path = PathBuf::from(format!(
+            "{}/{}_cxx.h",
```

Comment:
i wonder if `_cxx` is the best naming, as if you call your module (or the input file, see other comment above) something with cxx at the end it'll collide. Whereas if this was `{name}.cxx.h` or `{name}.rs.h` it can never collide as the name can't have a `.` ?

Reply:
Good point, `.cxx.{h/cpp}` would avoid that issue.